packetbeat/protos/amqp/amqp.go

```diff
@@ -0,0 +1,883 @@
+		offset = next
+	}
+	return false
+}
+
+func (amqp *Amqp) handleAmqp(m *AmqpMessage, tcptuple *common.TcpTuple, dir uint8) {
+	if amqp.mustHideCloseMethod(m) {
+		return
+	}
+	logp.Debug("amqp", "A message is ready to be handled")
+	m.TcpTuple = *tcptuple
+	m.Direction = dir
+	m.CmdlineTuple = procs.ProcWatcher.FindProcessesTuple(tcptuple.IpPort())
+
+	if m.Method == "basic.publish" {
+		amqp.handlePublishing(m)
+	} else if m.Method == "basic.deliver" || m.Method == "basic.return" ||
+		m.Method == "basic.get-ok" {
+		amqp.handleDelivering(m)
+	} else if m.IsRequest == true {
+		amqp.handleAmqpRequest(m)
+	} else if m.IsRequest == false {
+		amqp.handleAmqpResponse(m)
+	}
+}
+
+func (amqp *Amqp) mustHideCloseMethod(m *AmqpMessage) bool {
+	return amqp.Hide_connection_information == true &&
+		(m.Method == "connection.close" || m.Method == "channel.close") &&
+		getReplyCode(m.Fields) < uint16(300)
+}
+
+func (amqp *Amqp) handleAmqpRequest(msg *AmqpMessage) {
+	// Add it to the HT
+	tuple := msg.TcpTuple
+
+	trans := amqp.getTransaction(tuple.Hashable())
+	if trans != nil {
+		if trans.Amqp != nil {
+			logp.Debug("amqp", "Two requests without a Response. Dropping old request: %s", trans.Amqp)
+		}
+	} else {
+		trans = &AmqpTransaction{Type: "amqp", tuple: tuple}
+		amqp.transactions.Put(tuple.Hashable(), trans)
+	}
+
+	trans.ts = msg.Ts
+	trans.Ts = int64(trans.ts.UnixNano() / 1000)
+	trans.JsTs = msg.Ts
+	trans.Src = common.Endpoint{
+		Ip:   msg.TcpTuple.Src_ip.String(),
+		Port: msg.TcpTuple.Src_port,
+		Proc: string(msg.CmdlineTuple.Src),
+	}
+	trans.Dst = common.Endpoint{
+		Ip:   msg.TcpTuple.Dst_ip.String(),
+		Port: msg.TcpTuple.Dst_port,
+		Proc: string(msg.CmdlineTuple.Dst),
+	}
+	if msg.Direction == tcp.TcpDirectionReverse {
+		trans.Src, trans.Dst = trans.Dst, trans.Src
+	}
+
+	trans.Method = msg.Method
+	// get the righ request
+	if len(msg.Request) > 0 {
+		trans.Request = strings.Join([]string{msg.Method, msg.Request}, " ")
+	} else {
+		trans.Request = msg.Method
+	}
+	//length = message + 4 bytes header + frame end octet
+	trans.BytesIn = msg.Body_size + 12
+	if msg.Fields != nil {
+		trans.Amqp = msg.Fields
+	} else {
+		trans.Amqp = common.MapStr{}
+	}
+
+	//if error or exception, publish it now. sometimes client or server never send
+	//an ack message and the error is lost. Also, if nowait flag set, don't expect
+	//any response and publish
+	if isAsynchronous(trans) {
+		amqp.publishTransaction(trans)
+		logp.Debug("amqp", "Amqp transaction completed")
+		amqp.transactions.Delete(trans.tuple.Hashable())
+		return
+	}
+
+	if trans.timer != nil {
+		trans.timer.Stop()
+	}
+	trans.timer = time.AfterFunc(TransactionTimeout, func() { amqp.expireTransaction(trans) })
+}
+
+func (amqp *Amqp) handleAmqpResponse(msg *AmqpMessage) {
+	tuple := msg.TcpTuple
+	trans := amqp.getTransaction(tuple.Hashable())
+	if trans == nil || trans.Amqp == nil {
+		logp.Warn("Response from unknown transaction. Ignoring.")
+		return
```

Comment:
But responses being send are error responses, right? If so, I'd say these responses are even more interesting then the requests